util/client/ngmclient/client.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 PingCAP, Inc. Licensed under Apache-2.0.
+
+package ngmclient
+
+import (
+	"context"
+	"fmt"
+	"net/http/httputil"
+	"net/url"
+	"sync/atomic"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/joomcode/errorx"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/fx"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/pingcap/tidb-dashboard/pkg/utils/topology"
+	"github.com/pingcap/tidb-dashboard/util/topo"
+)
+
+var (
+	NgmErrNS                 = errorx.NewNamespace("error.api.ng_monitoring")
+	ErrNgMonitoringNotDeploy = NgmErrNS.NewType("not_deploy")
+	ErrNgMonitoringNotStart  = NgmErrNS.NewType("not_start")
+)
+
+const (
+	ngMonitoringCacheTTL = time.Second * 5
+)
+
+type ngMonitoringAddrCacheEntity struct {
+	address string
+	err     error
+	cacheAt time.Time
+}
+
+type NgmClient struct {
+	lifecycleCtx          context.Context
+	etcdClient            *clientv3.Client
+	ngMonitoringReqGroup  singleflight.Group
+	ngMonitoringAddrCache atomic.Value
+}
+
+func NewNgmClient(lc fx.Lifecycle, etcdClient *clientv3.Client) (*NgmClient, error) {
+	nc := &NgmClient{etcdClient: etcdClient}
+	lc.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			nc.lifecycleCtx = ctx
+			return nil
+		},
+	})
+
+	return nc, nil
+}
+
+func (nc *NgmClient) Route(targetPath string) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		ngMonitoringAddr, err := nc.getNgMonitoringAddrFromCache()
+		if err != nil {
+			_ = c.Error(err)
+			return
+		}
+
+		c.Request.URL.Path = targetPath
+
+		ngMonitoringURL, _ := url.Parse(ngMonitoringAddr)
+		proxy := httputil.NewSingleHostReverseProxy(ngMonitoringURL)
+		proxy.ServeHTTP(c.Writer, c.Request)
+	}
+}
+
+func (nc *NgmClient) getNgMonitoringAddrFromCache() (string, error) {
```

Comment:
It is never a good idea to resolve the address by the client itself. It means your client will have only one limited use case. Usually proving something small, atomic, cohesive and then assembling these things is a better solution. 

Think about cloud, and other dashboard variants. None of them can provide you a resolution in this way. It effectively let your util cannot be used in these variants.